e2e/test/scenarios/native/native.cy.spec.js

```diff
@@ -314,6 +319,65 @@ describe("scenarios > question > native", () => {
     cy.findByText("Showing 1 row").should("be.visible");
   });
 
+  describe("no native access", () => {
+    beforeEach(() => {
+      restore("postgres-12");
+      cy.signInAsAdmin();
+      cy.intercept("/api/database?saved=true").as("database");
+      cy.updatePermissionsGraph({
+        [USER_GROUPS.ALL_USERS_GROUP]: {
+          [WRITABLE_DB_ID]: { data: { schemas: "none", native: "none" } },
+        },
+        [USER_GROUPS.NOSQL_GROUP]: {
+          [SAMPLE_DB_ID]: { data: { schemas: "all", native: "write" } },
+          [WRITABLE_DB_ID]: { data: { schemas: "all", native: "none" } },
+        },
+      });
+
+      cy.updateCollectionGraph({
+        [USER_GROUPS.NOSQL_GROUP]: { root: "write" },
+      });
+
+      cy.createNativeQuestion(
+        {
+          name: "Secret Orders",
+          native: {
+            query: "SELECT * FROM ORDERS",
+          },
+          database: WRITABLE_DB_ID,
+        },
+        {
+          wrapId: true,
+        },
+      );
+
+      cy.signIn("nosql");
+    });
+
+    it("should not display the query when you do not have native access to the data source", () => {
+      cy.get("@questionId").then(questionId =>
+        cy.visit(`/question/${questionId}`),
+      );
+
+      cy.findByTestId("native-query-top-bar").within(() => {
+        cy.findByText("This question is written in SQL.").should("be.visible");
```

Comment:
Just to zoom out a bit:

I understand we want to prevent restricted users from seeing the query (the issue is even labeled ".Security related").
But we do that only in frontend.
User can still look the query up in network tab in browser developer tools.

Shouldn't we also enforce this restriction in backend?
If not, what's the point of enforcing it in frontend?

![image](https://github.com/metabase/metabase/assets/6830683/af4a5c51-70e9-4a82-8958-8d02471d350d)

Reply:
I agree that if the purpose of this is to obscure the query from the end user, this doesn't really accomplish the task. The text for the query can be found in the network bar, or by observing the DOM (the div just has `display:none`).

Reply:
I guess we have an answer: https://github.com/metabase/metabase/pull/34285#issuecomment-1750698098